File: Patterns/Command Pattern/Cpp/Command/GarageDoor/GarageDoorCloseCommand.cpp
#include "../ICommand.cpp"
#include "./GarageDoor.cpp"

class GarageDoorCloseCommand : public ICommand {
private:
    GarageDoor* garageDoor;
    int preview;

public:
    GarageDoorCloseCommand(GarageDoor* garageDoor) : garageDoor{garageDoor}, preview{garageDoor->OFF} {}

    void execute() override {
        preview = garageDoor->getState();
        garageDoor->down();
        garageDoor->lightOff();
        garageDoor->stop();
    }

    void undo() override {
        if (preview == garageDoor->ON) {
            garageDoor->up();
            garageDoor->lightOn();
            garageDoor->stop();
        }
    }

    string getCommandName() override {
        return "GarageDoorCloseCommand";
    }
};

File: Patterns/Command Pattern/CSharp/Command/GarageDoor/GarageDoorCloseCommand.cs
namespace Command.GarageDoor
{
    public class GarageDoorCloseCommand : ICommand
    {
        private GarageDoor garageDoor;
        private int preview;

        public GarageDoorCloseCommand(GarageDoor garageDoor)
        {
            preview = GarageDoor.OFF;
            this.garageDoor = garageDoor;
        }

        public void Execute()
        {
            preview = garageDoor.State;
            garageDoor.Up();
            garageDoor.LightOn();
            garageDoor.Stop();
        }

        public void Undo()
        {
            if (preview == GarageDoor.ON)
            {
                garageDoor.Down();
                garageDoor.LightOff();
                garageDoor.Stop();
            }
        }
    }
}